### DISCUSSION TITLE
Position and orientation of meshes

### DISCUSSION THREAD

**Discussion by dyl1n** [google-deepmind/mujoco#2171] [Asking for Help] [answered]
### Intro

Hi!

I am a graduate student at University of Strasbourg, I am learning to use MuJoCo.

### My setup

MuJoCo version: 3.2.4
OS: macOS Sonoma

### My question

I wonder how meshes associated to a _geom_ are positioned and orientated. There are explanations in the documentation but I am not sure if I understood them correctly. I experimented a bit with simple meshes but I was not able to clarify my understanding. So I will expose what I understood, and if you could tell me if I am correct or not will be very helpful.

In the documentation, one can read: _**“Positioning and orienting is complicated by the fact that vertex data in the source asset are often relative to coordinate frames whose origin is not inside the mesh. In contrast, MuJoCo expects the origin of a geom’s local frame to coincide with the geometric center of the shape. We resolve this discrepancy by pre-processing the mesh in the compiler, so that it is centered around (0,0,0) and its principal axes of inertia are the coordinate axes.“**_

The _**“coordinate frame(s) whose origin is not inside the mesh“**_ is the coordinate frame in which the vertices are expressed in the mesh file (.obj for instance), let us denote it by $\mathcal{F}_L$ (L for Local). The _**“geom’s local frame“**_ is the frame specified (with respect to the body frame $\mathcal{F}_B$ of the body to which the geom belongs) by the user in the MJCF file of our model. Let us denote it by $\mathcal{F}_g$. It is the _geom_’s position and orientation before compilation.

My understanding of _**“We resolve this discrepancy by pre-processing the mesh in the compiler, so that it is centered around (0,0,0) and its principal axes of inertia are the coordinate axes.“**_ is that with respect to $\mathcal{F}_L$, the geometric center and principal axes of inertia of the mesh is computed, leading to a new frame $\mathcal{F}_I$ (I for inertial). And this frame is then superposed to $\mathcal{F}_g$.

One can now, express positions of vertices in the body frame $\mathcal{F}_B$. To keep notations simple I use homogeneous transformation matrices (they contain the same information that positions vectors and quaternions).

Let $^LT_I$ be the transform from $\mathcal{F}_L$ to $\mathcal{F}_I$, this transform depends only on the mesh file and the information stored in **mjModel.mesh_pos** and **mjModel.mesh_quat** corresponds to the inverse of $^LT_I$.
Let $^BT_g$ be the transform from $\mathcal{F}_B$ to $\mathcal{F}_g$, this transform is provided by the user as attributes of the _geom_.
And more generally, let $^XT_Y$ be the transform from frame $\mathcal{F}_X$ to $\mathcal{F}_Y$.

Now, if $v_n$ is the $n$-th vertex of the mesh, and if its position (homogeneous coordinate) in the mesh file (so with respect to $\mathcal{F}_L$) is $^Lv_n$, according to my understanding the position of this vertex with respect to the body frame $\mathcal{F}_B$ is:

$$^Bv_n=^BT_L.^Lv_n=^BT_I.^IT_L.^Lv_n=^BT_g.(^LT_I)^{-1}.^Lv_n$$

And the _geom_’s position and orientation after compilation is $^BT_L=^BT_g.(^LT_I)^{-1}$, which is different from the position and orientation specified by the user in MJCF. This corresponds to the following part of the documentation: _**“Unlike all other geoms, the position and orientation of mesh geoms after compilation do not equal the settings of the corresponding attributes here (i.e attributes of the geom). Instead they are offset by the translation and rotation that were needed to center and align the mesh asset in its own coordinate frame.“**_

This message is quite long because I tried to be very precise so as to avoid confusion. Is my understanding right? If not, what is false about it?

Thank you

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by dyl1n** [comment]
A picture is worth a thousand words
![IMG_9285 2](https://github.com/user-attachments/assets/11c8e022-4552-4e78-acb3-07b9c8d0f5fd)

**Comment by quagla** [comment, marked as answer]
Hi, yes it's the standard transformation composed with T^-1 on the right, as you wrote.

See here https://github.com/google-deepmind/mujoco/blob/main/src/engine/engine_ray.c#L767 for example how the inverse mesh transformation is applied to the geom frame in Cartesian coordinates in order to get the Cartiesian position of the frame specified in the input file (which here is an implicit SDF surface).

Here https://github.com/google-deepmind/mujoco/blob/main/src/user/user_mesh.cc#L1522 you can find how the mesh is transformed during model compilation.

**Reply by dyl1n** [reply]
Thank you for answer! I'll take a look at the code

**Comment by yuvaltassa** [comment]
Relatedly, you can find the transformation applied by the compiler to the original asset data in `mjModel.mesh_{scale,pos,quat}`.

After you feel that you've understood everything, feel free to send a PR to improve this section of the docs.

**Reply by dyl1n** [reply]
I would be very happy to improve this section of the documentation. I noticed that the XML Reference part of the docs contains almost no math (i.e., no LaTeX formulas), unlike the Computation section. Is this intentional?

My initial doubts about the mesh transformations arose because everything was explained in natural language, which, while helpful, isn't always as precise as math for complex concepts. If possible, I’d like to add concise math notation to the docs by defining the frames and relationships via transformation matrices, as I did in my original post. I believe this could enhance clarity for readers who rely on precise mathematical notation.

Please let me know if something similar to my post (but much more condensed) would be appropriate in the asset/mesh (*) subsection of the documentation.

**Reply by yuvaltassa** [reply]
Ya go for it!

Do you know how to build the documentation locally? @kevinzakka showed me once but I forgot. (That too should be in the docs somewhere :slightly_smiling_face: )

**Reply by yuvaltassa** [reply]
Ah [found it](https://github.com/google-deepmind/mujoco/issues/470#issuecomment-1236393265)
